[libs] Move Syscomm To Metal IO Sockets

csegarragonz opened the pull request:
In this PR I move the syscomm library to use metal IO sockets. 

These sockets are non-blocking by default, and the only way to block on them is via a `poll` structure. Given that in some places we rely on blocking reads, I implement a `BlockingSocketStream` object that wraps around the blocking logic. Thus, `SocketStream` only implements non-blocking `try_read/try_read_exact`, whereas `BlockingSocketStream` only implements blocking calls `read/read_exact`. We can get a `BlockingSocketStream` from a `SocketStream` by calling `SocketStream::set_blocking()`.

csegarragonz committed:
[libs] F: Move syscomm To Use Non-Blocking mio Sockets

In this commit I re-factor the syscomm library to use non-blocking metal
IO sockets. This is a necessary step towards implementing poll-based
multiplexing of user VM (and potentially gateway) conections.

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[microvm] B: Use Non-Blocking Sockets

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[utils] E: Make nanvix-bench Use Blocking Sockets

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[utils] E: Make Nanvixd Use Blocking Sockets

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

csegarragonz committed:
[daemons] E: Make Gateway Socket In Linuxd Blocking

In this commit I re-factor the gateway stream in linuxd to be a
BlockingSocketStream.

Signed-off-by: Carlos Segarra <carlos@carlossegarra.com>

ppenna commented on Cargo.toml:
```diff
@@ -80,6 +80,7 @@ kvm-ioctls = "0.23.0"
 libc = "0.2.172"
 log = "0.4.27"
 microvm = { path = "src/microvm", default-features = false }
+mio = { version = "1", default-features = false }
```
Is it possible to specify the full version number so we can have more reproducible builds?

ppenna commented on src/libs/config/src/lib.rs:
```diff
@@ -65,6 +65,13 @@ pub mod syscomm {
+    ///
+    /// # Description
+    ///
+    /// Provides the maximum number of poll events we can buffer.
```
Could you add the rationale for this value?

ppenna commented on src/utils/nanvix-bench/src/main.rs:
```diff
@@ -538,7 +550,20 @@ impl Benchmark {
+                    Ok(()) => break,
+                    Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => continue,
+                    Err(e) => {
+                        return Err(anyhow::anyhow!(
```
Please log an error before failing.

csegarragonz commented on src/libs/syscomm/src/lib.rs:
```diff
@@ -195,12 +324,132 @@ impl SocketStream {
+                let mut events = Events::with_capacity(config::syscomm::MAX_NUM_POLL_EVENTS);
+
+                // Even after a poll wake-up the socket may still return WouldBlock.
+                loop {
```
On a spurious wake-up we immediately poll again (and sleep).

ppenna commented on src/libs/syscomm/src/lib.rs:
```diff
@@ -167,16 +250,62 @@ impl SocketStream {
+                    return Err(io::Error::new(ErrorKind::UnexpectedEof, "connection closed").into())
+                },
+                Ok(n) => total_read += n,
+                Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
```
@csegarragonz I agree. this can be a problem, unless you carefully handle the `WOuldBlock` case.

csegarragonz replied:
Makes sense. I have addressed this and made `try_read_exact` return the number of bytes it managed to read.

csegarragonz committed:
[daemons] E: Make Linuxd Handle Partial Reads

In this commit I extend linuxd to graecfully handle partial reads from
the non-blocking socket. We do so by adding a small buffer together with
the non-blocking socket stream where we can cache partial reads.

The buffer dynamically grows, because it will always be strictly smaller
than a message size.

csegarragonz commented on src/libs/syscomm/src/lib.rs:
```diff
@@ -133,24 +181,59 @@ impl SocketStream {
+                    })?;
+
+                Ok(BlockingSocketStream::Unix(stream, poll_instance))
+            },
```
This is a deliberate design choice, and is only used for sockets that need to be shared across worker threads where we cannot have a centralized polling structure.

csegarragonz commented on src/utils/nanvix-bench/src/main.rs:
```diff
@@ -513,7 +512,26 @@ impl Benchmark {
+                            "error reading from input VMM stream (error={e:?})"
+                        ));
+                    },
+                }
```
This is only used in a special benchmark that needs access to the underlying sockets.

csegarragonz commented on src/microvm/src/gateway.rs:
```diff
@@ -84,23 +89,34 @@ impl Gateway {
+        }
+        // Post-condition: partial_read_buffer is empty.
+
+        match self.stream.try_read_exact(&mut buf) {
```
damn, good catch copilot

csegarragonz commented on src/libs/syscomm/src/lib.rs:
```diff
@@ -167,17 +250,59 @@ impl SocketStream {
+        while total_read < buf.len() {
+            match self.try_read(&mut buf[total_read..]) {
+                Ok(0) => {
+                    return Err(io::Error::new(ErrorKind::UnexpectedEof, "connection closed").into())
```
I don't think this is true? It should return WouldBlock.

ppenna commented on src/daemons/linuxd/src/linuxd/mod.rs:
```diff
@@ -189,10 +183,23 @@ impl LinuxDaemon {
+                                        error!(
+                                            "error acquiring lock on user VM stream (error={e:?})"
+                                        );
+                                        continue;
```
Failing to notify the user VM about this may have undefined behavior. If we fail to obtain the lock, it is likely that we had a contention, we could sleep and retry. If we retried for more than X times, than we should simply kill that VM (instead of letting it hanging). I understand that this is out of the scope of this PR. Could you create an issue and link it here?

csegarragonz replied:
If I understand correctly, [`lock()`](https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock) will only return an error if the mutex is poisoned, which means that another thread panicked while holding the lock.

If this happens this means that something is going very wrong, so maybe we should just error-out? (It is unlikely that we will be able to clean this user VM anyway, neither).

csegarragonz replied:
Created a tracking issue: https://github.com/nanvix/nanvix/issues/787

ppenna commented on src/daemons/linuxd/src/worker_thread.rs:
```diff
@@ -842,44 +851,41 @@ impl WorkerThreadHandle {
+
+                // Read from the gateway thread.
+                let mut response_buf: [u8; ReadResponse::BUFFER_SIZE] =
+                    [0u8; ReadResponse::BUFFER_SIZE];
```
Please use the EOF helper function.

ppenna commented on src/libs/syscomm/src/lib.rs:
```diff
@@ -19,16 +32,14 @@ use ::std::{
+// Constants
+//==================================================================================================
+
+const BLOCKING_THREAD_TOKEN: Token = Token(0);
```
Could you add a comment with a rationale for this constant here?
